### PR TITLE
Homework 2

### PR DESCRIPTION
**Pull request by nikita-talalai**

### CODE REVIEW

File: hw2/task1/Tests.fs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿module hw2.Tests
```
Comment (DedSec256) [inline hw2/task1/Tests.fs:1]: Не хватает проверки пустого списка и списка с отрицательными числами.

File: hw2/task2/Program.fs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+type BinTree<'a> =
+    | Node of 'a * BinTree<'a> * BinTree<'a>
+    | Empty
+    
+let mapTree mapping tree =
+    let rec innerMap tree cont =
+        match tree with
+        | Node(x, l, r) ->
+            innerMap l (fun left -> innerMap r (fun right -> cont (Node(mapping x, left, right))))
+        | Empty -> cont Empty
```
Comment (DedSec256) [inline hw2/task2/Program.fs:12]: За старания 5ка, но в реальной жизни более простое решение "в лоб" без страшной хвостовой рекурсии с CPS здесь подошло бы лучше, вероятно.

File: hw2/task3/Program.fs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿module task3.Program
+
+type Expression =
+    | Const of int
+    | Plus of Expression * Expression
+    | Minus of Expression * Expression
+    | Times of Expression * Expression
+    | Divide of Expression * Expression
```
Comment (DedSec256) [inline hw2/task3/Program.fs:8]: Можно упростить до BinaryExpression

File: hw2/task4/Program.fs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿module task4.Program
+let isPrime n =
+    match n with
+    | n when n <= 1 -> false
+    | n when n = 2 -> true
+    | n ->
+        2 :: [3 .. 2 .. (n |> float |> sqrt |> int) + 1]
+        |> List.filter (fun x -> n % x = 0)
+        |> List.isEmpty
```
Comment (DedSec256) [inline hw2/task4/Program.fs:9]: лучше не аллоцировать целые списки для такой простой задачи

File: hw2/task4/Tests.fs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+let ``Edge cases are correct`` () =
+    (isPrime 0 && isPrime 1) |> should equal false
+
+[<Test>]
+let ``isPrime works correct for prime numbers`` () =
+    seq{2; 3; 5; 7; 11; 13; 17; 19; 23; 29} |> Seq.forall isPrime |> should equal true
+        
+[<Test>]
+let ``Expected generation`` () =
+    primeGenerator() |> Seq.skip 10 |> Seq.take 10 |> should equalSeq <| seq {31; 37; 41; 43; 47; 53; 59; 61; 67; 71}
```
Comment (DedSec256) [inline hw2/task4/Tests.fs:22]: самый полезный тест, остальные сравнивают функцию isPrime саму с собой, что некорректно.

**Review by DedSec256 — APPROVED** [review]
